### PR TITLE
[Feature]: return routed experts to reuse

### PR DESCRIPTION
**Pull request by RunningLeon**
## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

cache routed expert ids and return when finish seq

### pipeline
```python
from lmdeploy import pipeline, GenerationConfig, PytorchEngineConfig

if __name__ == '__main__':
    backend_config = PytorchEngineConfig(tp=1, enable_return_routed_experts=True)
    prompts = ['Hello who are you?' * 1024, 'Can you write a poem about AI?']  * 1
    gen_config = GenerationConfig(return_routed_experts=True)
    model_path = 'Qwen/Qwen3-30B-A3B'
    pipe = pipeline(model_path, backend_config=backend_config)
    reps = pipe(prompts, gen_config=gen_config)
    for res in reps:
        print(res)
```

### api server
1. Launch server
```shell
ray start --head --port 8345

RAY_ADDRESS='127.0.0.1:8345' \
lmdeploy serve api_server \
Qwen/Qwen3-30B-A3B \
--backend pytorch \
--server-port 24545 \
--log-level INFO \
--tp 2 \
--enable-return-routed-experts
```

2. call generate api

```python
import requests
import numpy as np
import ray
import base64


if __name__ == '__main__':
    ray_address = '127.0.0.1:8345' # same ray cluster 
    ray.init(address=ray_address)
    input_ids = [151644, 872, 198, 9707, 879, 525, 498, 30, 151645, 198, 151644, 77091, 198]
    prompt = None
    payload = dict(
        session_id = -1,
        prompt = prompt,
        input_ids = input_ids,
        max_tokens= 1024,
        return_routed_experts=True,
        include_stop_str_in_output=True,
    )
    port = 24545
    api_url=f"http://0.0.0.0:{port}/generate"
    headers = {'content-type': 'application/json'}
    response = requests.post(api_url, headers=headers, json=payload)
    assert response.status_code == 200
    output = response.json()
    meta_info = output['meta_info']
    if routed_experts := meta_info.get('routed_experts', None):
        if isinstance(routed_experts, str):
            data = base64.b64decode(routed_experts)
            routed_experts = ray.get(ray.cloudpickle.loads(data))
        else:
            routed_experts = np.asarray(routed_experts)
        print(routed_experts.shape)
```
## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
4. If the modification has a dependency on downstream projects of a newer version, this PR should be tested with all supported versions of downstream projects.
5. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: lmdeploy/pytorch/engine/executor/ray_executor.py [hunk 1]
```diff
@@ -351,6 +352,13 @@ def wakeup(self, tags: Optional[List[str]] = None):
             self.update_configs()
         self.collective_rpc('wakeup', (tags, ))
 
+    def serialize(self, obj) -> str:
+        """Serialize obj."""
+        ref = ray.put(obj)
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/executor/ray_executor.py:357]: Who manages the lifecycle of the object in ray?

File: lmdeploy/pytorch/engine/model_agent.py [hunk 2]
```diff
@@ -248,13 +248,14 @@ def model_forward(
...
             output = model(**input_dict)
-
+            if not isinstance(output, Dict):
+                output = dict(hidden_states=output)
             # InternVL-3.5-Flash will change the seqlen, model_metas during forward
             if context.model_metas is not None and context.model_metas[0] is not None:
                 model_metas = context.model_metas
-            seq_length = context.q_seqlens[:len(inputs.seq_length)]
-
-    return dict(hidden_states=output, model_metas=model_metas, seq_length=seq_length)
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/model_agent.py:257]: What's the reason to remove `hidden_states`?

File: lmdeploy/messages.py [hunk 3]
```diff
@@ -117,6 +117,9 @@ class GenerationConfig:
     preserve_cache: bool = False
     migration_request: Optional[MigrationRequest] = None
 
+    # router replay
+    return_routed_experts: bool = False
```
Comment (lvhan028) [inline lmdeploy/messages.py:121]: Shall we use `out_routed_experts` to maintain consistency with `out_logits` and `out_last_hidden_states`?

File: lmdeploy/pytorch/engine/executor/ray_executor.py (hunk 1 above)
Comment (RunningLeon) [reply lmdeploy/pytorch/engine/executor/ray_executor.py:357]: The caller should free the obj once call `ray.get(obj_ref)` by `ray._private.internal_api.free(obj_ref)`

File: lmdeploy/pytorch/engine/model_agent.py (hunk 2 above)
Comment (RunningLeon) [reply lmdeploy/pytorch/engine/model_agent.py:257]: Did not remove. To support multiple output tensors, the output returned by model would be Dict[str, Tensor] instead of Tensor.

File: lmdeploy/messages.py (hunk 3 above)
Comment (RunningLeon) [reply lmdeploy/messages.py:121]: For pipeline, `GenerationConfig` is often used by users, maybe we should keep consistent with openai protocal?

File: lmdeploy/pytorch/engine/engine.py [hunk 4]
```diff
@@ -896,6 +904,10 @@ def __need_logits(seqs: SeqList):
             """Need logits."""
             return any(seq.return_logits for seq in seqs)
 
+        def __need_routed_experts(seqs: SeqList):
```
Comment (grimoire) [inline lmdeploy/pytorch/engine/engine.py:907]: Why do we treat `return_routed_experts` as a sampling parameter?

File: lmdeploy/pytorch/engine/model_agent.py [hunk 5]
```diff
@@ -478,14 +493,23 @@ def gather(self, output):
 
             def get_output(self):
                 """Get tmp_output."""
-                if not return_logits:
+                if not (return_logits or return_routed_experts):
```
Comment (grimoire) [inline lmdeploy/pytorch/engine/model_agent.py]: refactor this code:

```python
if return_logits:
    out_hidden_states = ...
else:
    out_hidden_states = ...

if return routed_experts:
    routed_experts = ...
else:
    routed_experts = ...

return out_hidden_states, routed_experts

File: lmdeploy/pytorch/models/qwen3_moe.py [hunk 6]
```diff
@@ -441,6 +459,36 @@ def get_input_embeddings(self):
         """Get input embeddings."""
         return self.model.get_input_embeddings()
 
+    def make_buffers_cudagraph(self, graph_meta: CudaGraphMeta, input_ids, **kwargs):
+        """Make cudagraph buffers from forward inputs."""
+        max_tokens = graph_meta.max_tokens
+
+        input_buffers = super().make_buffers_cudagraph(graph_meta=graph_meta, **kwargs)
+        if self.enable_return_routed_experts:
+            input_buffers['all_routed_experts'] = input_ids.new_empty(
```
Comment (grimoire) [inline lmdeploy/pytorch/models/qwen3_moe.py]: Do we really need an input buffer?
Could we output it in forward so we can read it from output buffer

File: lmdeploy/pytorch/backends/cuda/graph_runner.py [hunk 7]
```diff
@@ -134,9 +124,8 @@ def forward(self, **kwargs):
         context = self.ctx_mgr.current_context()
         self.model.update_context_cudagraph(self.meta, context)
         self._graph.replay()
-
         output_buffers = self.meta.output_buffers
-        output = self.slice_output(output_buffers, kwargs)
+        output = self.model.get_outputs_cudagraph(output_buffers, **kwargs)
```
Comment (grimoire) [inline lmdeploy/pytorch/backends/cuda/graph_runner.py:128]: Should we update forward when  enable_graph=False so prefill and decoding can share same output format

File: lmdeploy/pytorch/models/qwen3_moe.py (hunk 6 above)
Comment (RunningLeon) [reply lmdeploy/pytorch/models/qwen3_moe.py]: No need to concat  outputs for each layer

File: lmdeploy/pytorch/engine/engine.py (hunk 4 above)
Comment (RunningLeon) [reply lmdeploy/pytorch/engine/engine.py:907]: similar to return logits, there may be cases when users only wants routed experts for some requests

File: lmdeploy/pytorch/models/qwen3_moe.py (hunk 6 above)
Comment (grimoire) [reply lmdeploy/pytorch/models/qwen3_moe.py]: I means why don't we create the buffer in forward

Comment (RunningLeon) [reply lmdeploy/pytorch/models/qwen3_moe.py]: OK

File: lmdeploy/pytorch/backends/cuda/graph_runner.py (hunk 7 above)
Comment (RunningLeon) [reply lmdeploy/pytorch/backends/cuda/graph_runner.py:128]: Done in line 213

**Review by grimoire — APPROVED** [review]
